tokio-util/src/net/mod.rs

```diff
@@ -0,0 +1,104 @@
+//! TCP/UDP/Unix helpers for tokio.
+
+use crate::either::Either;
+use std::future::Future;
+use std::io::Result;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+#[cfg(unix)]
+pub mod unix;
+
+/// A trait for a listener: `TcpListener` and `UnixListener`.
+pub trait Listener: Send + Unpin {
+    /// The stream's type of this listener.
+    type Io: tokio::io::AsyncRead + tokio::io::AsyncWrite;
+    /// The socket address type of this listener.
+    type Addr;
+
+    /// Polls to accept a new incoming connection to this listener.
+    fn poll_accept(&self, cx: &mut Context<'_>) -> Poll<Result<(Self::Io, Self::Addr)>>;
+
+    /// Accepts a new incoming connection from this listener.
+    fn accept(&self) -> ListenerAcceptFut<'_, Self>
```

Comment:
I think it's best to accept a `&mut self` here. Calling `accept` twice in parallel will result in one of them breaking due to lost wakeups.
```suggestion
    /// Polls to accept a new incoming connection to this listener.
    fn poll_accept(&mut self, cx: &mut Context<'_>) -> Poll<Result<(Self::Io, Self::Addr)>>;

    /// Accepts a new incoming connection from this listener.
    fn accept(&mut self) -> ListenerAcceptFut<'_, Self>
```

Reply:
That would make the API different from tokio's TCP listener and Unix listener. If they don't use &mut self there, why would I need it here?

Reply:
The `accept` fn in Tokio will not have the same lost wakeup issue as this one does because it isn't implemented by calling `poll_accept` on itself. However, we cant really replicate what Tokio does with a trait. The `&mut self` avoids the footgun.

Reply:
Ok I see!! Thanks for the explanation

Reply:
Fixed in fe231ec6